app/controllers/concerns/navigator_actions.rb

```diff
@@ -0,0 +1,156 @@
+# frozen_string_literal: true
+
+module NavigatorActions
+  extend ActiveSupport::Concern
+
+  private
+
+  def project_serializer(project)
+    {
+      id: project.code,
+      name: project.name,
+      url: project_path(project),
+      archived: project.archived,
+      has_children: project.has_children,
+      children_url: navigator_project_path(project),
+      children: []
+    }
+  end
+
+  def folder_serializer(folder)
+    {
+      id: "PF#{folder.id}",
+      name: folder.name,
+      url: projects_path(project_folder_id: folder.id),
+      archived: folder.archived,
+      icon: 'fas fa-folder',
+      has_children: folder.has_children,
+      children_url: navigator_project_folder_path(folder),
+      children: []
+    }
+  end
+
+  def experiment_serializer(experiment)
+    {
+      id: experiment.code,
+      name: experiment.name,
+      url: canvas_experiment_path(experiment),
+      archived: experiment.archived,
+      has_children: experiment.has_children,
+      children_url: navigator_experiment_path(experiment),
+      children: []
+    }
+  end
+
+  def my_module_serializer(my_module)
+    {
+      id: my_module.code,
+      name: my_module.name,
+      url: protocols_my_module_path(my_module),
+      archived: my_module.archived,
+      has_children: 0,
+      children: []
+    }
+  end
+
+  def load_projects(folder = nil, archived = false)
+    has_children = if !archived
+                     'SUM(CASE WHEN experiments.archived IS FALSE THEN 1 ELSE 0 END) > 0 AS has_children'
+                   else
+                     'SUM(CASE WHEN experiments.archived IS TRUE OR my_modules.archived IS TRUE
+                               THEN 1 ELSE 0 END) > 0 AS has_children'
+                   end
+    current_team.projects
+                .where(project_folder_id: folder)
+                .viewable_by_user(current_user, current_team)
+                .viewable_children_by_user(current_user)
+                .where('
+                  projects.archived = :archived OR
+                  (
+                    (
+                      experiments.archived = :archived OR
+                      my_modules.archived = :archived
+                    ) AND
+                    :archived IS TRUE
+                  )
+                ', archived: archived)
+                .select(
+                  'projects.id',
+                  'projects.name',
+                  'projects.archived',
+                  has_children
+                ).group('projects.id')
+  end
+
+  def load_folders(folder = nil, archived = false)
```

Comment:
the `load_` prefix is usually used when setting an instance variable (`@folders = ...`), so just call this `def project_folders`, same for the methods below